ceph-releases/luminous/ubuntu/16.04/daemon/Dockerfile

```diff
@@ -9,41 +9,36 @@ ENV CEPH_VERSION luminous
-
-# install prerequisites
-RUN DEBIAN_FRONTEND=noninteractive apt-get update && apt-get install -y wget unzip uuid-runtime python-setuptools udev dmsetup && \
-\
-# Install ceph, ganesha and etcd
-    wget -q -O- 'https://download.ceph.com/keys/release.asc' | apt-key add - && \
-    echo "deb http://download.ceph.com/debian-$CEPH_VERSION/ xenial main" | tee /etc/apt/sources.list.d/ceph-$CEPH_VERSION.list && \
-    echo "deb http://download.ceph.com/nfs-ganesha/deb-V2.5-stable/luminous/ xenial main" | tee /etc/apt/sources.list.d/nfs-ganesha.list && \
-    apt-get update && apt-get install -y --force-yes $PACKAGES && \
-    dpkg -s $PACKAGES && \
-    apt-get clean && rm -rf $PURGES && \
-\
-# Install confd
-    chmod +x /usr/local/bin/confd && mkdir -p /etc/confd/conf.d && mkdir -p /etc/confd/templates && \
-\
-# Install forego
-    wget -O /forego.tgz 'https://bin.equinox.io/c/ekMN3bCZFUn/forego-stable-linux-amd64.tgz' && \
-    cd /usr/local/bin && tar xfz /forego.tgz && chmod +x /usr/local/bin/forego && rm /forego.tgz
-
-# Install kubectl
-ADD https://storage.googleapis.com/kubernetes-release/release/${KUBECTL_VERSION}/bin/linux/amd64/kubectl /usr/local/bin/kubectl
-RUN chmod +x /usr/local/bin/kubectl
-
 # Add s3cfg file
 ADD s3cfg /root/.s3cfg
 
 # Add bootstrap script, ceph defaults key/values for KV store
 ADD *.sh ceph.defaults check_zombie_mons.py ./osd_scenarios/* entrypoint.sh.in disabled_scenario /
 
+# install prerequisites
+RUN DEBIAN_FRONTEND=noninteractive apt-get update && apt-get install -y wget unzip uuid-runtime python-setuptools udev dmsetup && \
+  \ # Install ceph, ganesha and etcd
+  wget -q -O- 'https://download.ceph.com/keys/release.asc' | apt-key add - && \
+  echo "deb http://download.ceph.com/debian-$CEPH_VERSION/ xenial main" | tee /etc/apt/sources.list.d/ceph-$CEPH_VERSION.list && \
+  echo "deb http://download.ceph.com/nfs-ganesha/deb-V2.5-stable/luminous/ xenial main" | tee /etc/apt/sources.list.d/nfs-ganesha.list && \
+  apt-get update && apt-get install -y  --no-install-recommends --force-yes $PACKAGES && \
+  dpkg -s $PACKAGES && \
+  apt-get clean && rm -rf $PURGES && \
+  \ # Download & install confd
+  wget -O /usr/local/bin/confd "https://github.com/kelseyhightower/confd/releases/download/v${CONFD_VERSION}/confd-${CONFD_VERSION}-linux-amd64" && \
+  chmod +x /usr/local/bin/confd && mkdir -p /etc/confd/conf.d && mkdir -p /etc/confd/templates && \
+  \ # Install forego
+  wget -O /forego.tgz 'https://bin.equinox.io/c/ekMN3bCZFUn/forego-stable-linux-amd64.tgz' && \
+  cd /usr/local/bin && tar xfz /forego.tgz && chmod +x /usr/local/bin/forego && rm /forego.tgz && \
+  \ # Install kubectl
+  wget -O /usr/local/bin/kubectl "https://storage.googleapis.com/kubernetes-release/release/${KUBECTL_VERSION}/bin/linux/amd64/kubectl" && \
+  chmod +x /usr/local/bin/kubectl && \
+  \ # Cleaning container
+  bash /clean_container.sh && rm /clean_container.sh
```

Comment:
Running `clean_container` here does not remove some of unneeded files from the base Ubuntu image. The base image is plain Ubuntu:16.04 and its going to bring in ~100MB of files even if you end up removing them in this step.

Rook solved this by slimming the base ubuntu image and exporting it to a base tarball and back in as a new base. See https://github.com/rook/rook/blob/master/images/base/Makefile. New docker daemon now support multi-stage builds which make this work much easier but its widely deployed yet.

Reply:
@bassam, just a FYI @ErwanAliasr1 has a commit for that. We tested it together, pushing it soon.